### ISSUE TITLE
Resizeable World Portal

### ISSUE THREAD

**Issue by Luke18033** [TeamSpen210/BEE2-items#574]
Here are three (3) screenshots to show what I mean:
![20160516113859_1](https://cloud.githubusercontent.com/assets/19228201/15301084/d3804dbe-1b60-11e6-9d8e-ca1c0dab520d.jpg)
I forgot to say in the screenshot that "Start Reversed" will simply put the portal facing outwards instead of inwards. There will be a rotation handle to decide which direction it rotates.
![20160516121548_1](https://cloud.githubusercontent.com/assets/19228201/15301091/d8d81f4e-1b60-11e6-81af-a5c3fdb8ec81.jpg)
In order for this to work, the button would probably have to be connected to both ends of the portal. There would also be a "start enabled" option to make the portal start on.
![20160516114101_1](https://cloud.githubusercontent.com/assets/19228201/15301095/dc4faebc-1b60-11e6-881e-7e14a317f00d.jpg)
If the screenshot above didn't really make it clear, the logic gates won't be part of the editor model either. I'm only using them so that I can have antilines running between the glass.

**Comment by TeamSpen210** [comment]
I could use glass like this. Three problems:

* Worldportals are very processor-intensive, and would likely cause FPS issues if you made them very large.
* If more than 4 portals are visible onscreen, the game's renderer glitches and all kinds of weirdness happens. This basically limits you to 1 portal in Coop, and maybe more in SP. That slot is taken up by the observation room in Old Aperture, so you're very limited. Careful geometry placement can avoid this, but that's not something I expect most users to do.
* Glass isn't just rectangular - wall blocks can be used to split it up, or make other strange shapes. Worldportals need to be rectangular, and wouldn't be placeable in these situations.

**Comment by Luke18033** [comment]
Another problem is the world portal light holes in P1.

**Comment by Luke18033** [comment]
That also explains why you can see through the walls in my Aperture Tag for Portal 1 maps, because they use a lot of worldportal lights. I'm probably going to update them with regular lights once Pr21 is out.

**Comment by Luke18033** [comment]
@TeamSpen210 Fixes:

1. Don't put worldportals into compiled chamber if they are too big.
2. Once again, don't put any worldportals into compiled chamber if there are too many in one area. If possible, you could only get rid of recently-placed portals.
3. Again, don't generate portals split apart by walls in the compiled chamber.

**Comment by BenVlodgi** [comment]
@Luke18033
> worldportal lights

Whut? Light is pre-calculated at compile time.

> Don't generate worldportals into the compiled chamber if...

Too many conditions to code around, even if they could be implemented, they might stop someone's puzzle that would have worked fine.
World portals can be turned on and off, what if someone has 10 of them linked together in a room, but only 1 set is ever on at a time based on the puzzle logic.

Fewer restrictions is always better.

**Comment by Luke18033** [comment]
@BenVlodgi "worldportal lights" refers to the worldportal light holes stylevar for P1. I used a ton of light holes in those maps, so the rendering broke.

**Comment by Luke18033** [comment]
@BenVlodgi Restrictions 1 and 3 should still apply, but restriction 2 can be removed. Maybe just put a warning in the item description that looking at too many activated worldportals breaks the rendering.

**Comment by BenVlodgi** [comment]
Why stop a world portal if its really large?
If the designer decides to bisect a wall with the portal, maybe stop them, but that isn't necessarily the worst thing ever.


I just think everyone should embrace the Hammer Master Race

**Comment by Luke18033** [comment]
Yeah, I guess.

**Comment by Luke18033** [comment]
So just 3 then?

**Comment by Luke18033** [comment]
Wait, what's the "hammer master race"?

**Comment by TeamSpen210** [comment]
Well, Hammer is better than anything PeTI can hope to achieve...

**Comment by Enterpise** [comment]
@BenVlodgi If you think that, then why you are wasting your time developing this mod? Instead develping this, you should make P2 authoring tool tutorials if you think everyone should join the hammer master race.

**Comment by Konclan** [comment]
Because dumbasses like me can't make the sendificator and such in hammer

**Comment by BenVlodgi** [comment]
@Enterpise when do you think the last time I made a code commit to this project was? BenVlodgi/BEE2.4@79c169ce273e5d0f7ab0cce4abcdb81e4e0fdf7a _(Sep 25, 2014)_ xD
Though it's not because I think it's waste of time, but I've slowly become unfamiliar with all the improvements and features @TeamSpen210 has made.
I've also been devoting more time to Backstock and other personal game projects.

> you should make P2 authoring tool tutorials if you think everyone should join the hammer master race.

I don't have the source code for that, although, I really do wish to complete a [pre-compiler](https://github.com/BenVlodgi/Backstock-Precompiler) that will collapse instances.

I'd love to have this conversation in the [chat](https://gitter.im/BenVlodgi/BEE2.4) though :) Maybe you could make some suggestions for improvements for Hammer, personally I'm not sure what I'd change if I had the chance.

**Comment by BenVlodgi** [comment]
@Konclan isn't it just an instance?

**Comment by Luke18033** [comment]
@BenVlodgi @Konclan @TeamSpen210 @Enterpise Alright, I didn't open an issue about resizeable worldportals just to have it turn into an argument that everyone should be using Hammer. I think I'm going to close this and copy the original comment to a new issue.

**Comment by BenVlodgi** [comment]
Leave this one open, no need to make more issues for the same thing. No-one had even posted here in a week

**Comment by Luke18033** [comment]
@BenVlodgi But I'd prefer if we talked less about Hammer and more about resizeable worldportals!

**Comment by Luke18033** [comment]
You know, maybe someday somebody will make an external editor tool which was like a combination of Hammer and the PeTI. I already have plans to make this in the future, but I don't want to make people forget about the BEE2. Not yet anyway.

**Comment by xDadiKx** [comment]
@Luke18033 That's my dream.
I want a easy editor with so many features from hammer that are made easily.

**Comment by Luke18033** [comment]
I have plans to make Puzzle Maker Plus or PeTI Plus in the future, which will be exactly that. It won't be done for a while though.

**Comment by xDadiKx** [comment]
Do you actually know how to do it? Like change the limit of items, add more items, styles?

**Comment by Luke18033** [comment]
Actually, I should have said that it won't be STARTED for a while. I'd most likely need experience with Hammer's compiler before I could make it. I have no idea how to use Hammer.

**Comment by xDadiKx** [comment]
My estimate is that you probably will start making around 2017. It does take a bit of time to learn hammer.

**Comment by Luke18033** [comment]
Luckily, I've found [a whole playlist of Hammer tutorials](https://www.youtube.com/playlist?list=PL-454Fe3dQH1L38FnKkz_O1CqYx6sKaXk) on YouTube.

**Comment by xDadiKx** [comment]
Thats what I watched. I will watch them again and relearn hammer again.

**Comment by Luke18033** [comment]
There's actually an [older set of hammer tutorials](https://www.youtube.com/playlist?list=PL-454Fe3dQH3ro-IplCis_8zGO1BYYYNR), which has more videos, but might not work in the latest version of Hammer/Source engine/whatever.

**Comment by Luke18033** [comment]
One feature of Puzzle Maker Plus I have planned is that it will be very easy to add items. You will be able to create custom properties for items, and expand the item palette to fit these items. Another feature will be the ability to have multiple styles in one chamber, so you can have a Clean test and escape into the BTS areas, before falling into a pit and ending up in 1980s Aperture, all in one map.

**Comment by TheDarkBomber** [comment]
@Luke18033 

And with stretchable triggers and instances, with logic items from Hammer as well!

**Comment by xDadiKx** [comment]
And Floating Triggers.

**Comment by Luke18033** [comment]
@TheDarkBomber @xDadiKx Those are actually two things I wanted to include. I still want to keep it simple though, so I'm probably going to still make it block-based, but make the blocks a lot smaller in the compiled chamber (So you can make stairs without the Stairs item etc). I also won't have things that could potentially cause leaks if used wrong. Some of this could be avoided, however, by creating custom items.

**Comment by Luke18033** [comment]
I also wanted to include resizeable worldportals (this issue) and player-only buttons (#457).

**Comment by Luke18033** [comment]
Btw one tile in the PeTI would be the same size as a 4x4 of tiles in Puzzle Maker Plus.

**Comment by TheDarkBomber** [comment]
With the Hammer section, would it be possible to include Adhesion Gel?

**Comment by Luke18033** [comment]
@TheDarkBomber Probably.

**Comment by Luke18033** [comment]
@TheDarkBomber @xDadiKx This issue is supposed to be about resizeable world portals. We should talk about this on Gitter instead.

**Comment by xDadiKx** [comment]
Then get online.

**Comment by TheDarkBomber** [comment]
So...

Puzzle Maker Plus. Welcome to the best discussion of Puzzle Maker Plus.

**Comment by Luke18033** [comment]
So, about the resizeable worldportals...are they being added?

**Comment by TheDarkBomber** [comment]
Well, there is actually a fizzler that is basically a "resizable worldportal", made by FourthReaper.

**Comment by Luke18033** [comment]
Yes, but, as I said on [Gitter](https://gitter.im/BenVlodgi/BEE2.4), I wanted a worldportal without a fizzler

**Comment by TheDarkBomber** [comment]
Hazard Type switches between the two.

**Comment by Luke18033** [comment]
@BenVlodgi I've finally listened to your advice about Hammer. It's actually pretty easy to start figuring stuff out once you start messing around.

**Comment by Luke18033** [comment]
So are resizable worldportals going to be added?

**Comment by TeamSpen210** [comment]
Yes, I think. One small thing is that I'll have to recreate the code which generates glass borders, since custom glass items don't generate them. I need that anyway though for breakable glass and other items.

**Comment by TheDarkBomber** [comment]
We could use the Surface Division Field.

**Comment by BenVlodgi** [comment]
Why do you need glass borders?
Just as an indicator?

**Comment by Luke18033** [comment]
@TheDarkBomber I DON'T WANT THE SURFACE DIVISION FIELD JUST A NORMAL WORLD PORTAL!

**Comment by Luke18033** [comment]
@TeamSpen210 Borders should only be generated on sides which aren't touching a wall, if you didn't do that than some maps might look strange.

**Comment by ZombiePwner1** [comment]
@Luke18033 A puzzle maker plus would be great but to change a few things like a built in style changer you'd need access to the games source code which we can't get I already asked @TeamSpen210 yesterday if it was possible and he said it wasn't

**Comment by Luke18033** [comment]
@ZombiePwner1 It will be a separate program.

**Comment by TheDarkBomber** [comment]
Technically it is possible to get the source code if you are a Valve Employee.

**Comment by Luke18033** [comment]
Or if a Valve employee gives it to you

**Comment by TheDarkBomber** [comment]
@Luke18033 That would probably only happen if you pay them £300 or something.

**Comment by BenVlodgi** [comment]
@Luke18033  @TheDarkBomber you can get the source code for $25,000

**Comment by Luke18033** [comment]
@TheDarkBomber We're not using the Surface Division Field. Stop asking for it.

**Comment by Luke18033** [comment]
> Borders should only be generated on sides which aren't touching a wall, if you didn't do that than some maps might look strange.

@TeamSpen210 Borders shouldn't be generated at all, actually.

**Comment by TheDarkBomber** [comment]
We can just use the **Surface Division Field** :+1: :100: :greenland:

**Comment by Endermage77** [comment]
Absolutely implement the SDF.
However, the problem with it is that the SDF is not a free standing object, unlike, say, this worldportal suggestion here.

**Comment by Byzarru** [comment]
@Endermage77  I've suggested before making it a standing fizzler. I'd love to see this item in action.
@BenVlodgi 
> you can get the source code for $25,000

Let's get this bake sale started!

**Comment by Endermage77** [comment]
I know! I LOVE non-Euclidean levels. It's my favorite way to go.
And to forward their spread, we must expand the worldportal empire.

**Comment by Luke18033** [comment]
The surface division field should be implemented too, but as a separate item. Since outputs don't work on fizzler items, we'd probably need some sort of "SDF Connector" item.

**Comment by Endermage77** [comment]
Is there some sort of problem with using the existing fizzler output item?

**Comment by XdplayerdX** [comment]
You can use that model (with the permission of the author):

https://www.youtube.com/watch?v=lshvflN2E3c

http://steamcommunity.com/sharedfiles/filedetails/?id=71642152

![download](https://cloud.githubusercontent.com/assets/25773302/26681934/2a1de732-46b5-11e7-82ae-8b84c1871031.jpg)

![download 1](https://cloud.githubusercontent.com/assets/25773302/26681941/2cfe1846-46b5-11e7-9ead-6427902f0a40.jpg)

**Comment by Luke18033** [comment]
Three reasons that model wouldn't work:
* The portal needs to be resizeable - if we used that model, it would look bad at larger sizes.
* That model is a circle, but the portal needs to be a square.
* The portal needs to be seamless, with no visible indication that there is a portal there (besides things like lighting changes).

**Comment by Luke18033** [comment]
In regards to the 4+ portals issue, I've found that setting `r_portal_fastpath` to `0` stops the bug. This might have some other unwanted effects though, especially on slower computers (every system I've tested this on has been a high-end gaming PC). We'd need to do some testing before being able to safely use this. Someone with source code access could also possibly be of some help, if the NDA doesn't get in the way.

**Comment by seagemgames** [comment]
Some COOL guy has made these as a UCP
https://github.com/benjaminpants/bsoa/releases/latest

**Comment by Luke18033** [comment]
Are those resizeable using glass logic, as I suggested here?

**Comment by seagemgames** [comment]
I don't know, I haven't actually used them. It'd be cool as a first party item, though.

**Comment by Luke18033** [comment]
I asked the P2CE devs about `r_portal_fastpath`. It can definitely affect performance, but we'd still need to test it on a bunch of different systems to figure out exactly by how much, and whether it's safe to use in maps.
